pkg/controller/operconfig/pod_controller.go

```diff
@@ -1,48 +1,80 @@
-
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 // newPodReconciler returns a new reconcile.Reconciler
-func newPodReconciler(status *statusmanager.StatusManager) *ReconcilePods {
-	return &ReconcilePods{status: status}
+// When an update is enqueued, it will call all the functions
+// provided, as long as the object's resourceVersion is new.
+// (This is needed because no-op updates are still broadcast
+// internally in controller-runtime.)
+func newPodReconciler(client client.Client, funcs ...func()) *ReconcilePods {
+	return &ReconcilePods{
+		client:      client,
+		updateFuncs: funcs,
+	}
 }
 
 var _ reconcile.Reconciler = &ReconcilePods{}
 
 // ReconcilePods watches for updates to specified resources and then updates its StatusManager
 type ReconcilePods struct {
-	status *statusmanager.StatusManager
-
-	resources []types.NamespacedName
+	client      client.Client
+	updateFuncs []func()
+	resources   map[types.NamespacedName]*uns.Unstructured
 }
 
-func (r *ReconcilePods) SetResources(resources []types.NamespacedName) {
-	r.resources = resources
+// SetResources tells this controller which Daemonsets and Deployments
+// we should care about. (This is because we have to watch all of them,
+// so we can filter updates).
+func (r *ReconcilePods) SetResources(resources []*uns.Unstructured) {
+	r.resources = map[types.NamespacedName]*uns.Unstructured{}
+	for _, obj := range resources {
+		r.resources[types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}] = obj
+	}
 }
 
-// Reconcile updates the ClusterOperator.Status to match the current state of the
-// watched Deployments/DaemonSets
+// Reconcile trigers any downstream subscribers to the deployment / daemonset
+// updates (usually StatusManager and the operconfig controller).
 func (r *ReconcilePods) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	found := false
-	for _, name := range r.resources {
-		if name.Namespace == request.Namespace && name.Name == request.Name {
-			found = true
-			break
-		}
+	nsn := types.NamespacedName{
+		Namespace: request.Namespace,
+		Name:      request.Name,
 	}
+
+	existing, found := r.resources[nsn]
 	if !found {
+		// We don't care about this object
 		return reconcile.Result{}, nil
 	}
 
-	log.Printf("Reconciling update to %s/%s\n", request.Namespace, request.Name)
-	r.status.SetFromPods()
+	newObj := (existing.NewEmptyInstance()).(*uns.Unstructured)
+	if err := r.client.Get(context.TODO(), nsn, newObj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("ReconcilePods: failed to retrieve object: %w", err)
+	}
+
+	if existing.GetResourceVersion() == newObj.GetResourceVersion() {
+		return reconcile.Result{}, nil
```

Comment:
This all seems weird and fragile (like we could end up in an infinite loop if there's a small bug in it)

Reply:
It is weird! There's definitely something wrong with the plumbing in `controller-runtime` where the PodReconciler is triggered even when it doesn't change on the server.

I started to look in to it, but hadn't gotten to the bottom of it.

I wish I had a better answer. We could also just set the reconciliation time to 2 minutes and just sweep this under the rug.